daemon/daemon.go

```diff
@@ -1515,6 +1515,34 @@ func CreateDaemonRoot(config *config.Config) error {
 	return setupDaemonRoot(config, realRoot, idMapping.RootPair())
 }
 
+// RemapContainerdNamespaces returns the right containerd namespaces to use:
+// - if they are not already set in the config file
+// -  and the daemon is running with user namespace remapping enabled
+// Then it will return new namespace names, otherwise it will return the existing
+// namespaces
+func RemapContainerdNamespaces(config *config.Config) (string, string, error) {
+	idMapping, err := setupRemappedRoot(config)
+	if err != nil {
+		return "", "", err
+	}
+	if idMapping.Empty() {
+		return config.ContainerdNamespace, config.ContainerdPluginNamespace, nil
+	}
+	root := idMapping.RootPair()
+
+	c8dns := config.ContainerdNamespace
+	if _, ok := config.ValuesSet["containerd-namespace"]; !ok {
```

Comment:
Dang, took me way too long to figure out why we access `config.ValuesSet` just to see if it's there, and return the base value if it's set, but return something new if it's not (the "new" value being a string that includes the "base" value). This entire "config*y*" part of the code is really messy to work with.

Reply:
Yeah I was confused by it at the start too, but good thing these things exist :)